semihosting/semihosting.rst

```diff
@@ -0,0 +1,567 @@
+
+3.	If the ``SYS_OPEN`` from step 1 did not fail, the caller must now use ``SYS_CLOSE`` to close the filehandle.
+
+It is important to follow this sequence to ensure correct behavior on legacy semihosting implementations, which do not recognize the special filename ``:semihosting-features``.
+
+
+Example pseudocode function for querying feature bits
+-----------------------------------------------------
+
+The following C-like pseudocode describes one possible simple implementation of a check for a specific feature bit:
+
+.. code-block:: none
+
+    #define MAGICLEN 4
+    bool sh_feature_supported(int bytenum, int bitnum)
+    {
+        unsigned char magic[MAGICLEN];
+        unsigned char c;
+        int fh;
+        int len;
+
+        fh = sys_open(":semihosting-features", 0);
+        if (fh == -1) {
+            return false;
+        }
+        len = sys_flen(fh);
+        if (len <= bytenum) {
+            sys_close(fh);
+            return false;
+        }
+        if (sys_read(fh, &magic, MAGICLEN) != 0) {
+            sys_close(fh);
+            return false;
+        }
+        if (magic[0] != SHFB_MAGIC_0 ||
+            magic[1] != SHFB_MAGIC_1 ||
+            magic[2] != SHFB_MAGIC_2 ||
+            magic[3] != SHFB_MAGIC_3) {
+            sys_close(fh);
+            return false;
+        }
+        if (sys_seek(fh, bytenum) != 0) {
+            sys_close(fh);
+            return false;
+        }
+        if (sys_read(fh, &c, 1) != 0) {
+            sys_close(fh);
+            return false;
+        }
+        sys_close(fh);
+        return (c & (1 << bitnum)) != 0;
+    }
+
+More sophisticated implementations can choose to cache the contents of the start of the file with the magic bytes and the first few bytes of feature bit data. Conversely if, for instance, the caller knows in advance that it is only interested in feature bits inside the first feature byte, it can simplify this method to perform a single 5 byte read and need not call ``SYS_FLEN`` or ``SYS_SEEK``.
+
+
+Semihosting extensions available
+--------------------------------
+
+The `Semihosting Extensions Table` lists the semihosting extensions currently available, along with their associated feature byte and bit.
+
+.. _Semihosting Extensions Table:
+
+.. table:: Semihosting Extensions
+
+  +-----------------------------+--------------+-------------+
+  | Name                        | Feature byte | Feature bit |
+  +=============================+==============+=============+
+  | `SH_EXT_EXIT_EXTENDED`_     |            0 |           0 |
+  +-----------------------------+--------------+-------------+
+  | `SH_EXT_STDOUT_STDERR`_     |            0 |           1 |
+  +-----------------------------+--------------+-------------+
+
+.. raw:: pdf
+
+   PageBreak
+
+Semihosting operations
+======================
+
+The following semihosting operations are available:
+
+.. include:: sys_clock_0x10.rst
```

Comment:
Unfortunately this isn't picked up properly by the web interface. Could you perhaps paste the different docs into this main doc?